CollateralAccount/CollateralAccount.md

```diff
@@ -0,0 +1,106 @@
+---
+CIP: ?
+Title: Collateral Account for HD Wallets
+Authors: Sebastien Guillemot <seba@dcspark.io>
+Comments-URI: https://forum.cardano.org/t/collateral-account-derivation/65879
+Status: Draft
+Type: Standards
+Created: 2021-06-29
+License: CC-BY-4.0
+---
+
+# Abstract
+
+This document describes using a separate derivation path to ensure there is always sufficient collateral to execute UTXO smart contracts
+
+# Motivation
+
+Collateral input | As of Alonzo, transactions that call non-native smart contracts are required to put up collateral to cover the potential cost of smart contract execution failure. Inputs used as collateral has the following properties:
+
+- Cannot contain any tokens (only ADA)
+- Cannot be a script address
+- Must be a UTXO input
+- Must be at least some percentage of the fee in the tx (concrete percentage decided by a protocol parameter)
+- Can be the same UTXO entry as used in regular tx input
+- Is consumed entirely (no change) if the contract execution fails during phase 2 validation
+
+We therefore need a way for the wallet to handle picking which inputs to use for collateral
+
+## Naive solution
+
+The naive solution would do the following steps
+1. Add any regular input that satisfies the constraints as collateral as well
+1. If collateral is still insufficient, look at the wallet UTXO for any other entry that can be added as collateral
+1. If there are not enough UTXO entries that can be used as collateral, check if we can refactor the wallet UTXO to create enough UTXO entries that match the requirement
+1. If refactoring can unblock the wallet, prompt the user to send a transaction to refactor their UTXO (and pay the tx fee associated with it) before sending their transaction. If refactoring can't unblock the wallet, tell the user they will need more ADA to call this smart contract
+
+This however, causes the following problems:
+1. This may cause users to risk more collateral than they are comfortable with. In general, software should run smart contracts locally to detect if a transaction would fail before sending and alert the user to avoid consuming the collateral, but relying on this is not ideal.
```

Comment:
When we were discussing collateral inputs, we were thinking that there might indeed be only one collateral UTxO, and that most transactions would be over-collateralised. I don't think relying on this is particularly risky. Firstly, we expect that most transactions spending from script-locked addresses will be submitted by the PAB, and hence already verified. Secondly, with https://github.com/input-output-hk/ouroboros-network/pull/3230 we expect the local node to reject submitted transactions that fail phase-2 validation.

Reply:
Unless you are running your own node, there is no guarantee anybody is setting this flag.

Reply:
I don't follow. 

It's quite simple: if your wallet is working correctly then you will never ever have any collateral taken. The local node that you submit your transaction two is just an extra double-check on this. So it should be really very safe indeed.

Reply:
Would make sense, have to describe the best practices, like you're mentioning about running your own node?

Reply:
There is no reason why a good-faith user should *ever* end up submitting a transaction with failing scripts. So the only people who are ever at risk of losing their collateral are malicious users or people who've inexplicably decided to go to some trouble to not check their transactions before sending them.

Reply:
What about light wallets or hosted APIs, where the operator of the service does not want to consume its computational resources but instead wants to delegate this responsibility to the user?

Reply:
I've lost my fair share of collateral on the Alonzo Blue testnet :)

Can we really categorically rule out that wallets and/or PAB will never send transactions with failing scripts? Should the user pay with a large collateral sum in case of a bug?

Reply:
The testnet node currently accepts transactions with failing scripts and posts them to the network as failures. That will be fixed before mainnet.